src/rpc/rawtransaction.cpp

```diff
@@ -755,30 +756,18 @@ UniValue signrawtransaction(const JSONRPCRequest& request)
     const CKeyStore& keystore = tempKeystore;
 #endif
 
-    int nHashType = SIGHASH_ALL | SIGHASH_FORKID;
+    SigHashType nHashType = SigHashType::ALL | SigHashType::FORKID;
     if (request.params.size() > 3 && !request.params[3].isNull()) {
-        static map<string, int> mapSigHashValues =
-            boost::assign::map_list_of
-            (std::string("ALL"), int(SIGHASH_ALL))
-            (std::string("ALL|ANYONECANPAY"), int(SIGHASH_ALL | SIGHASH_ANYONECANPAY))
-            (std::string("ALL|FORKID"), int(SIGHASH_ALL | SIGHASH_FORKID))
-            (std::string("ALL|FORKID|ANYONECANPAY"), int(SIGHASH_ALL | SIGHASH_FORKID | SIGHASH_ANYONECANPAY))
-            (std::string("NONE"), int(SIGHASH_NONE))
-            (std::string("NONE|ANYONECANPAY"), int(SIGHASH_NONE | SIGHASH_ANYONECANPAY))
-            (std::string("NONE|FORKID"), int(SIGHASH_NONE | SIGHASH_FORKID))
-            (std::string("NONE|FORKID|ANYONECANPAY"), int(SIGHASH_NONE | SIGHASH_FORKID | SIGHASH_ANYONECANPAY))
-            (std::string("SINGLE"), int(SIGHASH_SINGLE))
-            (std::string("SINGLE|ANYONECANPAY"), int(SIGHASH_SINGLE | SIGHASH_ANYONECANPAY))
-            (std::string("SINGLE|FORKID"), int(SIGHASH_SINGLE | SIGHASH_FORKID))
-            (std::string("SINGLE|FORKID|ANYONECANPAY"), int(SIGHASH_SINGLE | SIGHASH_FORKID | SIGHASH_ANYONECANPAY));
         string strHashType = request.params[3].get_str();
-        if (mapSigHashValues.count(strHashType))
-            nHashType = mapSigHashValues[strHashType];
-        else
-            throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid sighash param");
+        try {
+            nHashType = FromStr(strHashType);
```

Comment:
Do the other implementations have a dependency on flag ordering as this does?

Reply:
Looks like BU does not have a dependency on ordering, but will also not tell you when you mess up (passing ALL|NONE magically converts into SINGLE).


 https://github.com/BitcoinUnlimited/BitcoinUnlimited/blob/release/src/rpc/rawtransaction.cpp#L820

Reply:
Ouch this is not good for BU. Thanks for the report @dagurval!

Reply:
@sickpig I looked again and missed a detail, you have a `=` and not `|=` for the base flags.

This is even worse, because flag ordering is required but not enforced. As it stands `FORKID|SINGLE` results in `SINGLE`. `FORKID|ANYONECANPAY|SINGLE|NONE` results in `NONE`. `NOFORKID|FORKID` results in a sighashtype with FORKID, but signs it the wrong way.